File: ui-tests/helpers/apiHelpers.ts
import axios from 'axios';
import { generateRandomUser } from './testData';

export async function createRandomUserViaAPI() {
  const user = generateRandomUser();

  const response = await axios.post('http://localhost:3000/api/users', {
    user: {
      username: user.username,
      email: user.email,
      password: user.password,
    }
  });

  return {
    username: response.data.user.username,
    email: response.data.user.email,
    password: user.password,
    token: response.data.user.token,
  };
}


export async function createArticleViaAPI(token: string, article: any) {
  const response = await axios.post(
    'http://localhost:3000/api/articles',
    { article },
    { headers: { Authorization: `Token ${token}` } }
  );

  return {
    ...article,
    slug: response.data.article.slug,
  };
}

export async function addCommentViaAPI(token: string, slug: string, commentData: { body: string }) {
  const response = await fetch(`http://localhost:3000/api/articles/${slug}/comments`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Token ${token}`,
    },
    body: JSON.stringify({ comment: commentData }),
  });

  if (!response.ok) {
    throw new Error(`Failed to add comment: ${response.status} ${response.statusText}`);
  }

  return await response.json();
}

File: mobile-tests/helpers/api_helpers.py
# helpers/api_helpers.py

import requests
from helpers.test_data import generate_random_user
import os

# Get API host from environment variable, default to localhost for local development
API_HOST = os.getenv("API_HOST", "localhost")
API_PORT = os.getenv("API_PORT", "3000")
API_URL = f"http://{API_HOST}:{API_PORT}/api"

def create_random_user_via_api():
    user = generate_random_user()

    response = requests.post(
        f"{API_URL}/users",
        json={"user": {
            "username": user["username"],
            "email": user["email"],
            "password": user["password"]
        }}
    )

    response.raise_for_status()
    data = response.json()

    return {
        "username": data["user"]["username"],
        "email": data["user"]["email"],
        "password": user["password"],
        "token": data["user"]["token"]
    }

def create_article_via_api(token, article):
    headers = {"Authorization": f"Token {token}"}
    response = requests.post(
        f"{API_URL}/articles",
        json={"article": article},
        headers=headers
    )
    response.raise_for_status()
    data = response.json()

    return {
        **article,
        "slug": data["article"]["slug"]
    }

def add_comment_via_api(token, slug, comment_body):
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Token {token}"
    }
    response = requests.post(
        f"{API_URL}/articles/{slug}/comments",
        json={"comment": {"body": comment_body}},
        headers=headers
    )

    if response.status_code != 200:
        raise Exception(f"Failed to add comment: {response.status_code} {response.text}")

    return response.json()
